substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -1987,6 +2155,74 @@ impl<T: Config> Pallet<T> {
 		Self::ensure_disabled_validators_sorted()
 	}
 
+	/// Invariants:
+	/// * If the election preparation has started (i.e. `now`  >= `expected_election - n_pages`):
+	///     * The election preparation metadata should be set (`ElectingStartedAt`);
+	///     * The electable stashes should not be empty;
+	///     * The exposures for the current electable stashes should have been collected;
+	/// * If the election preparation has not started yet:
+	///     * The election preparation metadata is empty;
+	///     * The electable stashes for this era is empty;
+	pub fn ensure_snapshot_metadata_state(now: BlockNumberFor<T>) -> Result<(), TryRuntimeError> {
+		let pages: BlockNumberFor<T> = Self::election_pages().into();
+		let next_election = <Self as ElectionDataProvider>::next_election_prediction(now);
+		let expect_election_start_at = next_election.saturating_sub(pages);
+
+		let election_prep_started = now >= expect_election_start_at;
+
+		// check election metadata, electable targets and era exposures if election should have
+		// already started.
+		match election_prep_started {
```

Comment:
It would probably be simpler to read if this was an `if !election_prep_started { ... return Ok(()); }` and then followed by the rest.